### PR TITLE
SLING-11160 : Repoinit does not allow to remove individual ACEs (jcr impl)

### PR DESCRIPTION
**Pull request by anchela**
@bdelacretaz , @cziegeler , @karlpauls , implementation of the REMOVE of a dedicated ACE which was not supported up to now. i would appreciate if you had time to review it.

NOTE: build will fail unless PR to repoinit.parser is merged. so for the time being this is for illustration of the problem with the existing repo-init language and why i suggest to change it).

### CODE REVIEW

File: src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java [hunk 1]
```diff
@@ -692,28 +693,64 @@ public void testRemoveMatchingEntry() throws Exception {
...
     @Test
     public void testRemoveNoMatchingEntry() throws Exception {
+        Principal principal = getPrincipal(U.username);
         String setup = "set principal ACL for " + U.username + "\n"
                 + "allow jcr:write on "+path+"\n"
                 + "end";
         U.parseAndExecute(setup);
+        assertPolicy(principal, U.adminSession, 1);
 
-        setup = "set principal ACL for " + U.username + "\n"
```
Comment (bdelacretaz) [inline src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java:702]: IIUC this test goes away with your changes, shouldn't it be kept? To test backwards compatibility - even if the corresponding operation is deprecated.

File: src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java [hunk 2]
```diff
@@ -692,28 +693,64 @@ public void testRemoveMatchingEntry() throws Exception {
...
-        } catch(RuntimeException rex) {
-            assertRegex(REMOVE_NOT_SUPPORTED_REGEX, rex.getMessage());
-        }      
+        assertPolicy(principal, U.adminSession, 1);
     }
 
-    @Test(expected = RuntimeException.class)
+    @Test(expected = RepoInitException.class)
     public void testRemoveNonExistingPrincipal() throws Exception {
-        String setup = "set principal ACL for nonExistingPrincipal\n"
```
Comment (bdelacretaz) [inline src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java:715]: Same comment as above for a test that goes away

File: src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java [hunk 3]
```diff
@@ -725,16 +762,16 @@ public void testRemovePrincipalMismatch() throws Exception {
...
+        assertPolicy(getPrincipal(U.username), U.adminSession, 1);
 
-        try {
-            setup = "set principal ACL for otherSystemPrincipal\n"
-            + "remove jcr:write on " + path + "\n"
-            + "end";
-            U.parseAndExecute(setup);
-            fail("Expecting REMOVE to fail");
-        } catch(RuntimeException rex) {
-            assertRegex(REMOVE_NOT_SUPPORTED_REGEX, rex.getMessage());
```
Comment (bdelacretaz) [inline src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java:736]: Same comment as above for a test that goes away

File: src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java (hunk 2 above)
Comment (anchela) [reply src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java:715]: this one didn't go away.... instead i refactored the tests. i think 1 test to make sure the unsupported remove-action really throws an exception is sufficient.... instead wanted to have extra coverage for the new removal.

File: src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java (hunk 3 above)
Comment (anchela) [reply src/test/java/org/apache/sling/jcr/repoinit/PrincipalBasedAclTest.java:736]: see above.

**Comment by anchela** [comment]
@bdelacretaz , i renamed the methods/classes as discussed and would appreciate if you could take another look

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 4]
```diff
@@ -291,6 +328,42 @@ public static void setPrincipalAcl(Session session, String principalName, Collec
...
+            Privilege[] privs = AccessControlUtils.privilegesFromNames(session, privNames.toArray(new String[0]));
+            Predicate<PrincipalAccessControlList.Entry> predicate = entry -> {
+                if (!jcrPaths.contains(entry.getEffectivePath())) {
+                    return false;
+                }
+                LocalAccessControlEntry lace = new LocalAccessControlEntry(entry.getPrincipal(), privs, line.getAction()== AclLine.Action.ALLOW, restr);
+                return lace.isEqual(entry);
+            };
+            if (removePrincipalEntries(acl, principalName, predicate)) {
+                modified = true;
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:357]: Looking at the test coverage with a `-P jacoco-report` build reports that this line is not covered by tests, and that's also the case for the `if (modified)` statement below. Those lines look important, I suppose they should be tested.

**Review by bdelacretaz — CHANGES_REQUESTED** [review]
The PR looks good to me but a test seems to be missing, see the comment around `modified = true;`. Once that's fixed I'm fine with merging.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 4 above)
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:357]: yes, i agree

Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:357]: @bdelacretaz , in fact there was a test-method for that but i managed to garble the test during the many refactorings. should be fixed now. can you please confirm?

**Review by bdelacretaz — APPROVED** [review]

**Pull request merged**